qutip/core/cy/cqobjevo.pxd

```diff
@@ -37,15 +37,23 @@ cimport numpy as cnp
 from qutip.core.data.base cimport idxint
 from qutip.core.data cimport CSR, Dense, Data
 
+
+cpdef enum LTYPE:
+    MIXED_TYPE   = 0
+    CSR_TYPE     = 1
+    Dense_TYPE   = 2
+
+
 cdef class CQobjEvo:
     cdef readonly (idxint, idxint) shape
-    cdef object dims
+    cdef readonly object dims
```

Comment:
This gives a false sense of safety - while `readonly` prevents you from reassigning the `list`, since the type is mutable you can still just change it anyway and mess stuff up.  That said, it's not like we can _really_ stop anyone in Python from doing whatever the hell the like, if they really put their mind to it.

Reply:
They can change them at the Qobj and QobjEvo level anyway. This could be solved later when we use more than list for dims.

Reply:
Ok, I'm happy with this then.  `readonly` is correct in principle, and the new `dims` will be immutable (I'm pretty sure).